src/source/custom_source.js

```diff
@@ -0,0 +1,339 @@
+ * They can trigger rendering using {@link Map#triggerRepaint}.
+ *
+ * @interface CustomSourceInterface
+ * @property {string} id A unique source id.
+ * @property {string} type The source's type. Must be `"custom"`.
+ * @example
+ * // Custom source implemented as ES6 class
+ * class CustomSource {
+ *     constructor() {
+ *         this.id = 'custom-source';
+ *         this.type = 'custom';
+ *         this.tileSize = 256;
+ *         this.tilesUrl = 'https://stamen-tiles.a.ssl.fastly.net/watercolor/{z}/{x}/{y}.jpg';
+ *         this.attribution = 'Map tiles by Stamen Design, under CC BY 3.0';
+ *     }
+ *
+ *     async loadTile(tile, {signal}) {
+ *         const url = this.tilesUrl
+ *             .replace('{z}', String(tile.z))
+ *             .replace('{x}', String(tile.x))
+ *             .replace('{y}', String(tile.y));
+ *
+ *         const response = await fetch(url, {signal});
+ *         const data = await response.arrayBuffer();
+ *
+ *         const blob = new window.Blob([new Uint8Array(data)], {type: 'image/png'});
+ *         const imageBitmap = await window.createImageBitmap(blob);
+ *
+ *         return imageBitmap;
+ *     }
+ * }
+ *
+ * map.on('load', () => {
+ *     map.addSource('custom-source', new CustomSource());
+ *     map.addLayer({
+ *         id: 'layer',
+ *         type: 'raster',
+ *         source: 'custom-source'
+ *     });
+ * });
+ */
+
+/**
+ * Optional method called when the source has been added to the Map with {@link Map#addSource}.
+ * This gives the source a chance to initialize resources and register event listeners.
+ *
+ * @function
+ * @memberof CustomSourceInterface
+ * @instance
+ * @name onAdd
+ * @param {Map} map The Map this custom source was just added to.
+ */
+
+/**
+ * Optional method called when the source has been removed from the Map with {@link Map#removeSource}.
+ * This gives the source a chance to clean up resources and event listeners.
+ *
+ * @function
+ * @memberof CustomSourceInterface
+ * @instance
+ * @name onRemove
+ * @param {Map} map The Map this custom source was added to.
+ */
+
+/**
+ * Optional method called after the tile is unloaded from the map viewport. This
+ * gives the source a chance to clean up resources and event listeners.
+ *
+ * @function
+ * @memberof CustomSourceInterface
+ * @instance
+ * @name unloadTile
+ * @param {{ z: number, x: number, y: number }} tile Tile name to unload in the XYZ scheme format.
+ * @param {Function} callback A callback to be called when the tile is unloaded.
+ */
+
+/**
+ * Optional method called if the map canceled request in loadTile. This
+ * gives the source a chance to clean up resources and event listeners.
+ *
+ * @function
+ * @memberof CustomSourceInterface
+ * @instance
+ * @name abortTile
+ * @param {{ z: number, x: number, y: number }} tile Tile name that was aborted in the XYZ scheme format.
+ * @param {Function} callback A callback to be called when the tile is aborted.
```

Comment:
Similar to `unloadTile`, I think we can remove the callback from this interface and keep it as `abortTile({z: number, x: number, y: number})`

Reply:
As we discussed, we decided to drop the `abortTile` in favor of AbortSignal in `loadTile`